README.md
# SimpCache
**SimpCache** is a caching library for Flutter that allows you to store and retrieve objects in a simple and efficient manner. The library creates a local variable of type `Map` and stores each object as `JSON` within it. Additionally, it also stores the `JSON` representation of each object in SharedPreferences using the `EncryptedSharedPreferences` package to provide an extra layer of security.

When you retrieve an object from **SimpCache**, it first checks the local variable to see if the object is already available. If it is not, it checks the SharedPreferences and retrieves the object as a `JSON` file. This approach allows **SimpCache** to provide fast and efficient access to cached objects.

To use **SimpCache** with complex classes, you must implement the `toJson` and `fromJson` methods for those classes. These methods are used to serialize and deserialize the objects into `JSON` format, which is then stored and retrieved by **SimpCache**. It's important to note that this requirement is due to the lack of Dart Mirroring/Reflect in Flutter.

Overall, **SimpCache** provides a simple and efficient caching solution for Flutter developers, and its use of SharedPreferences and encryption make it a secure choice for storing sensitive data.

## Features 👇
- 🔸 Efficient caching: SimpCache provides an efficient caching solution for Flutter apps by storing objects in a local variable and SharedPreferences, which makes accessing cached objects fast and efficient.

- 🔸 Secure data storage: SimpCache uses the `EncryptedSharedPreferences` package to store JSON data in SharedPreferences, which provides an extra layer of security for sensitive data.

- 🔸 Easy to use: SimpCache is easy to use and requires minimal configuration. Once you have implemented the `toJson` and `fromJson` methods for your complex classes, you can start caching objects with just a few lines of code.

- 🔸 Works with complex classes: SimpCache allows you to cache complex objects that don't have a simple data type, by serializing and deserializing them to and from JSON format.


## Installing ✅
```yaml
dependencies:
  simp_cache: ^<latest_version>
```

## Usage ⚒️
##### *Create an instance of SimpCache*
```
final simp = SimpCache.instance;
```

1. Cache objects
##### *Cache objects using simp.cacheItem*
```swift
final Student student = Student(  
  id: 1,  
  name: 'Yazan',  
  marks: [  
    Marks(  
      english: [9, 30, 42],  
	  arabic: [7, 28, 37],  
   ),  
    Marks(  
      english: [10, 36, 45],  
	  arabic: [5, 31, 40],  
   ),  
  ],  
);  
  
await simp.cacheItem(student.id.toString(), student);
```

>Note : Student Class must implement toJson & fromJson methods

example:
```swift
  class Student {  
 late int id;  
 late String name;  
 late List<Marks> marks;  
  
  Student({required this.id, required this.name, required this.marks});  
  
  Student.fromJson(Map<String, dynamic> json) {  
    id = json['id'];  
  name = json['name'];  
 if (json['marks'] != null) {  
      marks = <Marks>[];  
  json['marks'].forEach((v) {  
        marks.add(new Marks.fromJson(v));  
  });  
  }  
  }  
  
  Map<String, dynamic> toJson() {  
    final Map<String, dynamic> data = new Map<String, dynamic>();  
  data['id'] = this.id;  
  data['name'] = this.name;  
  data['marks'] = this.marks.map((v) => v.toJson()).toList();  
 return data;  
  }  
}  
  
class Marks {  
  List<int>? english;  
  List<int>? arabic;  
  
  Marks({this.english, this.arabic});  
  
  Marks.fromJson(Map<String, dynamic> json) {  
    english = json['English'].cast<int>();  
  arabic = json['Arabic'].cast<int>();  
  }  
  
  Map<String, dynamic> toJson() {  
    final Map<String, dynamic> data = new Map<String, dynamic>();  
  data['English'] = this.english;  
  data['Arabic'] = this.arabic;  
 return data;  
  }  
}
```
2. Retrieve objects as JSON
##### *Use simp.getItemByKey*
```swift
//showSource -> Shows if the item is retrieved from RAM or SharedPreferences  
final result = await simp.getItemByKey('1', showSource: true);  
  
//Convert JSON to Student class  
final Student student = Student.fromJson(jsonDecode(result));  
  
debugPrint(student.name);  
  
//OUTPUT : Yazan
```
3. Show cached objects

##### *Use simp.showCachedItems*

```swift
//If kDebugMode it will print all cached items.  
await simp.showCachedItems(showSource: true);
```
4. Remove cached object
##### *Use simp.removeItemByKey*
```swift
await simp.removeItemByKey('1');
```
5. Clear all cached objects
##### *Use simp.clearCache*
```swift
await simp.clearCache();
```
## Bugs/Requests 🔴
Feel free to reach out and open an issue if you encounter any problems.
If you feel the library is missing a feature, please raise a ticket on Github and I'll look into it.
Pull request are also welcome.



CHANGELOG.md
## 0.0.2

* Removed warning and formatted `simp_cache.dart`.


lib/simp_cache.dart
library simp_cache;

import 'dart:convert';

import 'package:encrypted_shared_preferences/encrypted_shared_preferences.dart';
import 'package:flutter/foundation.dart';
import 'package:simp_cache/src/simp_cache_constants.dart';

class SimpCache {
  SimpCache._internal();
  static final SimpCache instance = SimpCache._internal();

  Map<String, dynamic> _cachedItems = {};

  ///This method caches item in SharedPreferences and returns a boolean to indicate success or failure.
  Future<bool> cacheItem(String key, Object value) async {
    final EncryptedSharedPreferences _encryptedSharedPreferences =
        EncryptedSharedPreferences();
    final _secured = await _encryptedSharedPreferences.getInstance();
    _cachedItems[key] = jsonEncode(value);
    return _secured.setString(
        SimpCacheConstants.items, jsonEncode(_cachedItems));
  }

  ///This method is Future and returns a cached item if it exists.
  Future<String> getItemByKey(String key, {bool showSource = false}) async {
    await _getCachedItems(showSource: showSource);
    return _cachedItems.containsKey(key) ? _cachedItems[key] : '';
  }

  ///This method removes items from Ram & SharedPreferences and returns a boolean to indicate success or failure.
  Future<bool> removeItemByKey(String key) async {
    final EncryptedSharedPreferences _encryptedSharedPreferences =
        EncryptedSharedPreferences();
    final _secured = await _encryptedSharedPreferences.getInstance();
    await _getCachedItems();
    _cachedItems.remove(key);
    return _secured.setString(
        SimpCacheConstants.items, jsonEncode(_cachedItems));
  }

  ///This method prints all cached items (in debug mode only).
  Future<void> showCachedItems({bool showSource = false}) async {
    await _getCachedItems(showSource: showSource);
    _cachedItems.forEach((key, value) {
      if (kDebugMode) {
        debugPrint('$key - $value');
      }
    });
  }

  ///This method is not Future and returns a cached item by checking RAM only, if the item doesn't exist then it returns null.
  String? getItemByKeySync(String key) {
    return _cachedItems.containsKey(key) ? _cachedItems[key] : null;
  }

  ///This method clears all cached items and empties RAM then returns a boolean to indicate success or failure.
  Future<bool> clearCache() async {
    _cachedItems = {};

    final EncryptedSharedPreferences _encryptedSharedPreferences =
        EncryptedSharedPreferences();
    final _secured = await _encryptedSharedPreferences.getInstance();
    return _secured.remove(SimpCacheConstants.items);
  }

  /// ---------------- \\\

  Future<Map<String, dynamic>?> _getCachedItems(
      {bool showSource = false}) async {
    final EncryptedSharedPreferences _encryptedSharedPreferences =
        EncryptedSharedPreferences();
    final _secured = await _encryptedSharedPreferences.getInstance();

    try {
      if (_cachedItems.isNotEmpty) {
        if (showSource && kDebugMode) {
          debugPrint('SimpCache from RAM');
        }
        return _cachedItems;
      } else if (_secured.containsKey(SimpCacheConstants.items)) {
        if (showSource && kDebugMode) {
          debugPrint('SimpCache from SharedPreferences');
        }
        _cachedItems = await _getItemsFromDeviceStorage();
        _secured.setString(SimpCacheConstants.items, jsonEncode(_cachedItems));

        return _cachedItems;
      } else {
        return null;
      }
    } catch (error) {
      debugPrint('SimpCache error @ : $error');
      return null;
    }
  }

  Future<Map<String, dynamic>> _getItemsFromDeviceStorage() async {
    final EncryptedSharedPreferences _encryptedSharedPreferences =
        EncryptedSharedPreferences();
    final _secured = await _encryptedSharedPreferences.getInstance();
    final value = _secured.getString(SimpCacheConstants.items) ?? '';

    return jsonDecode(value);
  }
}


example/lib/student.dart
class Student {
  late int id;
  late String name;
  late List<Marks> marks;

  Student({required this.id, required this.name, required this.marks});

  Student.fromJson(Map<String, dynamic> json) {
    id = json['id'];
    name = json['name'];
    if (json['marks'] != null) {
      marks = <Marks>[];
      json['marks'].forEach((v) {
        marks.add(new Marks.fromJson(v));
      });
    }
  }

  Map<String, dynamic> toJson() {
    final Map<String, dynamic> data = new Map<String, dynamic>();
    data['id'] = this.id;
    data['name'] = this.name;
    data['marks'] = this.marks.map((v) => v.toJson()).toList();
    return data;
  }
}

class Marks {
  List<int>? english;
  List<int>? arabic;

  Marks({this.english, this.arabic});

  Marks.fromJson(Map<String, dynamic> json) {
    english = json['English'].cast<int>();
    arabic = json['Arabic'].cast<int>();
  }

  Map<String, dynamic> toJson() {
    final Map<String, dynamic> data = new Map<String, dynamic>();
    data['English'] = this.english;
    data['Arabic'] = this.arabic;
    return data;
  }
}

example/lib/main.dart
import 'dart:convert';

import 'package:example/student.dart';
import 'package:flutter/foundation.dart';
import 'package:flutter/material.dart';
import 'package:simp_cache/simp_cache.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({super.key, required this.title});

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  final simp = SimpCache.instance;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            TextButton(
              onPressed: onCacheItemTap,
              child: const Text(
                'Cache Item!',
              ),
            ),
            TextButton(
              onPressed: onGetItemTap,
              child: const Text(
                'Get Item By Key!',
              ),
            ),
            TextButton(
              onPressed: onRemoveItemTap,
              child: const Text(
                'Remove Item By Key!',
              ),
            ),
            TextButton(
              onPressed: onShowCacheTap,
              child: const Text(
                'Show Cached Items!',
              ),
            ),
            TextButton(
              onPressed: onGetItemSyncTap,
              child: const Text(
                'Get Item By Key Sync!',
              ),
            ),
            TextButton(
              onPressed: onClearCacheTap,
              child: const Text(
                'Clear Cache!',
              ),
            ),
          ],
        ),
      ),
    );
  }

  Future<void> onCacheItemTap() async {
    final Student student = Student(
      id: 1,
      name: 'Yazan',
      marks: [
        Marks(
          english: [9, 30, 42],
          arabic: [7, 28, 37],
        ),
        Marks(
          english: [10, 36, 45],
          arabic: [5, 31, 40],
        ),
      ],
    );

    await simp.cacheItem(student.id.toString(), student);
  }

  Future<void> onGetItemTap() async {
    //showSource -> Shows if the item is retrieved from RAM or SharedPreferences
    final result = await simp.getItemByKey('1', showSource: true);

    //Convert JSON to Student class
    final Student student = Student.fromJson(jsonDecode(result));

    debugPrint(student.name);

    //OUTPUT : Yazan
  }

  Future<void> onShowCacheTap() async {
    //If kDebugMode it will print all cached items.
    await simp.showCachedItems(showSource: true);
  }

  Future<void> onRemoveItemTap() async {
    await simp.removeItemByKey('1');
  }

  Future<void> onGetItemSyncTap() async {
    //Checks RAM only (Not Future).
    simp.getItemByKeySync('1');
  }

  Future<void> onClearCacheTap() async {
    await simp.clearCache();
  }
}
